### PR TITLE
Adding support for terraform provider registry

### PR DESCRIPTION
**Pull request by sponte**
Thanks for creating citizen - a missing registry for Terraform Modules. I have a need to host custom providers and thought of extending your project.

This PR adds support for hosting custom providers utilising Hashicorp's provider registry protocol documented at https://www.terraform.io/docs/internals/provider-registry-protocol.html.

## Changes

1. Added new endpoint for providers
2. Added new endpoint for publishers
3. Added `citizen` cli commands for publishing providers and trusted publishers
4. Added unit/integration tests that cover new functionality
5. Updated README to reflect the changes

I have updated the version to `1.0.0` to indicate breaking changes:

1. `citizen publish` command is no longer top level command as `citizen` can now publish modules, providers and publishers
2. Internals: Changed storage paths for modules to accommodate providers, now the root of the storage folder will contain top level `modules` and `providers` directories. This means new version won't work with existing registries, although careful migration is possible

### CODE REVIEW

File: lib/server.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-const http = require('http');
+const https = require('https');
```
Comment (ad-m) [inline lib/server.js:1]: It looks like out-of-scope changes.

Comment (sponte) [reply lib/server.js:2]: Yeah, will probably remove it but terraform provider registry requires HTTPS - once I've sorted the tests out, I should be able to remove https and https-localhost

Comment (sponte) [reply lib/server.js]: I have now removed `https-localhost` dependency as I added tests that utilise `ngrok`

**Comment by sponte** [comment]
Ok @ad-m - this is ready to review now - please have a look and let me know if you need me to make any changes before merging.

**Comment by sponte** [comment]
Not sure who will be reviewing this @outsideris - I referenced @ad-m earlier as he commented on the commit before but since you approved my other PRs (HCLv2/Terraform v12 and small update to the docs), I thought I'd bring this one to your attention. Please let me know your thoughts.

**Review by ad-m — APPROVED** [review]

**Comment by ad-m** [comment]
For clarity, I have no relationship with @outsideris . I am part of the community of this project because I follow it. I never even ddownloaded and started it.

**Comment by outsideris** [comment]
Sorry for late response. 
I will check your contribution soon. 
Thank you.

File: .gitignore [hunk 2]
```diff
@@ -5,3 +5,10 @@ node_modules/
 dist/
 test/integration/temp/
 data/
+.binaries/
+test/tf
+.terraform/
+.env
+test/integration/fixture/provider
+coverage/
+.nyc_output/
```
Comment (outsideris) [inline .gitignore:14]: I will clean up gitignore.

File: lib/server.js [hunk 3]
```diff
@@ -19,11 +19,12 @@ const normalizePort = (val) => {
   return false;
 };
 
-const run = () => {
+const run = async () => {
```
Comment (outsideris) [inline lib/server.js:22]: we don't need `async` here.

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "citizen",
-  "version": "0.3.3",
+  "version": "1.0.0",
```
Comment (outsideris) [inline package.json:3]: I don't want to version up now.

File: routes/providers.js [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+  form.on('close', async () => {
+    try {
+      // const exist = await hasProvider(`${destPath}/${filename}`);
+      // if (exist) {
+      //   const error = new Error('Provider exist');
+      //   error.status = 409;
+      //   error.message = `${destPath} is already exist.`;
+      //   return next(error);
+      // }
```
Comment (outsideris) [inline routes/providers.js:171]: unnecessary comments

**Review by outsideris — APPROVED** [review]
Thank you for your contribution. Now we support providers!!! 

Because I refactored code base on master after this PR, I merged this PR manually on my local. And then I will test and modify citizen intensively. (So, you don't need to fix this PR.)

**Comment by outsideris** [comment]
I rebased this PR on my local, see https://github.com/outsideris/citizen/commit/d698abfc135af318d47276e6246c97923e0702ee

File: package.json (hunk 4 above)
Comment (sponte) [reply package.json:3]: The reason version went up is that this MR introduces breaking changes in citizen cli

**Pull request closed without merging**